packages/worker/src/worker.ts

```diff
@@ -768,16 +831,59 @@ export class Worker {
-                      workflowType,
-                      workflowId,
-                      activation.runId,
+                      {
+                        filename: workflowType,
+                        runId: activation.runId,
+                        workflowId,
+                        namespace: this.options.serverOptions.namespace,
+                        taskQueue: this.options.taskQueue,
+                        isReplaying: activation.isReplaying,
+                      },
                       randomnessSeed,
-                      this.options.taskQueue
+                      this.options.isolateExecutionTimeoutMs
                     );
                   });
                   workflow = createdWF;
-                  // TODO: this probably shouldn't be here, consider alternative implementation
-                  await instrument(span, 'workflow.inject.console', () => createdWF.inject('console.log', console.log));
+                  await instrument(span, 'workflow.inject.dependencies', async () => {
+                    await createdWF.injectGlobal(
+                      'console.log',
+                      (...args: any[]) => {
+                        if (createdWF.info.isReplaying) return;
+                        console.log(
+                          `${createdWF.info.filename} (${createdWF.info.workflowId}-${createdWF.info.runId}) >`,
+                          ...args
+                        );
+                      },
+                      ApplyMode.SYNC
+                    );
+
+                    if (includesDeps(this.options)) {
+                      for (const [ifaceName, dep] of Object.entries(this.options.dependencies)) {
+                        for (const [fnName, impl] of Object.entries(dep)) {
+                          await createdWF.injectDependency(
+                            ifaceName,
+                            fnName,
+                            (...args) => {
+                              if (!impl.callDuringReplay && createdWF.info.isReplaying) return;
+                              try {
+                                const ret = impl.fn(createdWF.info, ...args);
+                                if (ret instanceof Promise) {
+                                  return ret.catch((error) =>
+                                    this.handleExternalDependencyError(createdWF.info, impl.applyMode, error)
+                                  );
+                                }
+                                return ret;
+                              } catch (error) {
+                                this.handleExternalDependencyError(createdWF.info, impl.applyMode, error);
+                              }
+                            },
+                            impl.applyMode,
+                            getIvmTransferOptions(impl)
+                          );
+                        }
+                      }
+                    }
+                  });
```

Comment:
Good candidate to separate into its own function

Reply:
Yeah, this got longer than I expected, will extract.